### PR TITLE
#1: Added schema converter to map data types to Spark schema

### LINKED ISSUE #1: Data types to Spark Schema mapping
## Situation

One of the main common requirements is to map Exasol datatypes, give a query or a table, into Spark schema. We should add this feature into the common library.

This is usually called `schema inference` and performed before reading or writing the data. 

For example, given the user query `SELECT * FROM database.table`, we should be able to infer a schema for Apache Spark by sending a modified query to get small dataset and then create a Spark schema from the returned query result set.

### Acceptance Criteria

- Added schema inference
- Exasol data types are mapped to Spark data types
- Unit and integration tests are added

### References

- Exasol data types: https://docs.exasol.com/db/latest/sql_references/data_types/datatypesoverview.htm

### Depends On

Blocked by #2

### PR DESCRIPTION
**Pull request by morazow**
Fixes #1

### CODE REVIEW

**Comment by morazow** [comment]
Sonar analysis fails, but should be fine for now, it should pick up default fall back branch once merged into main.

File: src/main/java/com/exasol/spark/common/SchemaConverter.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    /**
+     * Converts a list of column descriptions to Spark {@link StructType}.
+     *
+     * @param columns list of column descriptions
+     * @return Spark schema
+     */
+    public StructType convert(final List<ColumnDescription> columns) {
+        if (columns == null || columns.isEmpty()) {
+            throw new IllegalArgumentException(ExaError.messageBuilder("E-SCCJ-3")
+                    .message("Provided list of column descriptions are empty or null.")
```
Comment (kaklakariada) [inline src/main/java/com/exasol/spark/common/SchemaConverter.java:29]:
```suggestion
                    .message("Provided list of column descriptions is empty or null.")
```

File: src/main/java/com/exasol/spark/common/SchemaConverter.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+                    .mitigation("Please make sure that table or query has column definitions.").toString());
+        }
+        final List<StructField> fields = columns.stream().map(this::convertColumn).collect(Collectors.toList());
+        return new StructType(fields.toArray(new StructField[0]));
```
Comment (kaklakariada) [inline src/main/java/com/exasol/spark/common/SchemaConverter.java:33]: Use `Stream.toArray(...)` to directly create an array

File: src/main/java/com/exasol/spark/common/SchemaConverter.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+    private DataType createDecimalDataType(int precision, int scale) {
+        if (precision > DecimalType.MAX_PRECISION()) {
+            throw new IllegalStateException(ExaError.messageBuilder("E-SCCJ-5") //
+                    .message("Exasol decimal precision {{exasolPrecision}} is larger than "
+                            + "maximum Spark decimal precision {{sparkPrecision}}.") //
+                    .parameter("exasolPrecision", precision) //
+                    .parameter("sparkPrecision", DecimalType.MAX_PRECISION()) //
+                    .ticketMitigation() //
```
Comment (kaklakariada) [inline src/main/java/com/exasol/spark/common/SchemaConverter.java:100]:
```suggestion
```
We can't do anything there. Please add a note to the error message that this is not supported.

File: src/main/java/com/exasol/spark/common/SchemaConverter.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+                    .toString());
+        }
+        if (scale > DecimalType.MAX_SCALE()) {
+            throw new IllegalStateException(ExaError.messageBuilder("E-SCCJ-6") //
+                    .message("Exasol decimal scale {{exasolScale}} is larger than "
+                            + "maximum Spark decimal scale {{sparkScale}}.") //
+                    .parameter("exasolScale", scale) //
+                    .parameter("sparkScale", DecimalType.MAX_SCALE()) //
+                    .ticketMitigation() //
```
Comment (kaklakariada) [inline src/main/java/com/exasol/spark/common/SchemaConverter.java:109]:
```suggestion
```
We can't do anything there. Please add a note to the error message that this is not supported.

File: src/main/java/com/exasol/spark/common/SchemaConverter.java [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            return DataTypes.ShortType;
+        case Types.DOUBLE:
+        case Types.FLOAT:
+            return DataTypes.DoubleType;
+        case Types.REAL:
+            return DataTypes.FloatType;
+        case Types.INTEGER:
+            return column.isSigned() ? DataTypes.IntegerType : DataTypes.LongType;
+        case Types.BIGINT:
+            return column.isSigned() ? DataTypes.LongType : DataTypes.createDecimalType(20, 0);
```
[the remark this answers is not in the document]
Comment (morazow) [reply src/main/java/com/exasol/spark/common/SchemaConverter.java:71]: I have added short message, although it is not applicable to Exasol types

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**